h2o-core/src/main/java/water/rapids/Merge.java

```diff
@@ -340,7 +341,121 @@ public void map(Chunk[] cs, NewChunk[] ncs) {
+        SingleThreadRadixOrder.OXHeader oxheader = DKV.getGet(k);
+        DKV.remove(k);
+        if (oxheader != null) {
+          for (int b=0; b<oxheader._nBatch; ++b) {
+            k = SplitByMSBLocal.getSortedOXbatchKey(isLeft!=0, msb, b);
+            DKV.remove(k);
+          }
+        }
+      }
+    }
+    System.out.println("took: " + (System.nanoTime() - t0)/1e9);
+
+    System.out.print("Allocating and populating chunk info (e.g. size and batch number) ...");
+    t0 = System.nanoTime();
+    long ansN = 0;
+    int numChunks = 0;
+    for( SortCombine thisbm : bmList )
+      if( thisbm._numRowsInResult > 0 ) {
+        numChunks += thisbm._chunkSizes.length;
+        ansN += thisbm._numRowsInResult;
+      }
+    long chunkSizes[] = new long[numChunks];
+    int chunkLeftMSB[] = new int[numChunks];  // using too much space repeating the same value here, but, limited
+    int chunkRightMSB[] = new int[numChunks]; // leave it alone so as not to re-write chunkStitcher.
+    int chunkBatch[] = new int[numChunks];
+    int k = 0;
+    for( SortCombine thisbm : bmList ) {
+      if (thisbm._numRowsInResult == 0) continue;
+      int thisChunkSizes[] = thisbm._chunkSizes;
+      for (int j=0; j<thisChunkSizes.length; j++) {
+        chunkSizes[k] = thisChunkSizes[j];
+        chunkLeftMSB [k] = thisbm._leftSB._msb;
+        chunkRightMSB[k] = -1;
+        chunkBatch[k] = j;
+        k++;
+      }
+    }
+    System.out.println("took: " + (System.nanoTime() - t0) / 1e9);
+
+    // Now we can stitch together the final frame from the raw chunks that were
+    // put into the store
+    System.out.print("Allocating and populated espc ...");
+    t0 = System.nanoTime();
+    long espc[] = new long[chunkSizes.length+1];
+    int i=0;
+    long sum=0;
+    for (long s : chunkSizes) {
+      espc[i++] = sum;
+      sum+=s;
+    }
+    espc[espc.length-1] = sum;
+    System.out.println("took: " + (System.nanoTime() - t0) / 1e9);
+    assert(sum==ansN);
+
+    System.out.print("Allocating dummy vecs/chunks of the final frame ...");
+    t0 = System.nanoTime();
+    int numLeftCols = leftFrame.numCols();
+    int numColsInResult = numLeftCols;
+    final byte[] types = new byte[numColsInResult];
+    final String[][] doms = new String[numColsInResult][];
+    final String[] names = new String[numColsInResult];
+    for (int j=0; j<numLeftCols; j++) {
+      types[j] = leftFrame.vec(j).get_type();
+      doms[j] = leftFrame.domains()[j];
+      names[j] = leftFrame.names()[j];
+    }
+    
+    Key<Vec> key = Vec.newKey();
+    Vec[] vecs = new Vec(key, Vec.ESPC.rowLayout(key, espc)).makeCons(numColsInResult, 0, doms, types);
+    System.out.println("took: " + (System.nanoTime() - t0) / 1e9);
+
+    System.out.print("Finally stitch together by overwriting dummies ...");
+    t0 = System.nanoTime();
+    Frame fr = new Frame(names, vecs);
+    ChunkStitcher ff = new ChunkStitcher(chunkSizes, chunkLeftMSB, chunkRightMSB, chunkBatch);
```

Comment:
ChunkStitcher will stitch together Chunks formed for all MSBs into one big frame.